mbq/client/contrib/permissions.py

```diff
@@ -449,23 +459,43 @@ def has_all_permissions(  # noqa: F811
         )
         return result
 
+    @staticmethod
+    def _keyfn(e):
+        if e.startswith("company"):
+            return "company"
+        if e.startswith("vendor"):
+            return "vendor"
+        return "uuid"
+
+    def _parse_raw_org_refs(self, raw_org_refs: Iterable[str]) -> ConvenientOrgRefs:
+        grouped_org_refs = groupby(
+            sorted(raw_org_refs, key=self._keyfn), key=self._keyfn
+        )
+        company_ids, vendor_ids, org_refs = set(), set(), set()
+        for k, vs in grouped_org_refs:
+            if k == "company":
+                company_ids = {int(ref.split(":")[1]) for ref in vs}
+            if k == "vendor":
+                vendor_ids = {int(ref.split(":")[1]) for ref in vs}
+            else:
+                org_refs = set(vs)
+
+        return ConvenientOrgRefs(org_refs, company_ids, vendor_ids)
+
```

Comment:
I know I was the one who talked about `itertools.groupby` yesterday, but I think the way this implementation shook out is changing my mind. I thought it could have been a bit more declarative than this, and maybe that means it needs to be more imperative instead. In particular, the fact that we process each group of "raw refs" differently means that maybe the grouping idiom is not a good fit.

```python3
# more imperative, only one iteration over `raw_org_refs`
for raw_ref in raw_org_refs:
    if raw_ref.startswith("company"):
        company_ids.add(raw_ref.split(":")[1])
    ...
```

Reply:
@jreut I agree with you here. There is little value here using a `groupby` and forces us to provide more overhead to make it work with this use case: the `_keyfn` function is the indicator here since it is doing the same work as the `if` statements in the main function.